runtime/src/identity.rs

```diff
@@ -589,6 +637,106 @@ decl_module! {
+
+
+        /// It adds signing keys to target identity `id`.
+        /// Keys are directly added to identity because each of them has an authorization.
+        ///
+        /// Arguments:
+        ///     - `origin` Master key of `id` identity.
+        ///     - `id` Identity where new signing keys will be added.
+        ///     - `additional_keys` New signing items (and their authorization data) to add to target
+        ///     identity.
+        ///
+        /// Failure
+        ///     - It can only called by master key owner.
+        ///     - Keys should be able to linked to any identity.
+        pub fn add_signing_items_with_authorization( origin,
+                id: IdentityId,
+                expires_at: T::Moment,
+                additional_keys: Vec<SigningItemWithAuth>) -> Result {
+            let sender = ensure_signed(origin)?;
+            let sender_key = Key::try_from(sender.encode())?;
+            let _grants_checked = Self::grant_check_only_master_key(&sender_key, id)?;
+
+            // 0. Check expiration
+            let now = <timestamp::Module<T>>::get();
+            ensure!( now < expires_at, "Offchain authorization has expired");
+            let authorization = TargetIdAuthorization {
+                target_id: id,
+                nonce: Self::offchain_authorization_nonce(id),
+                expires_at
+            };
+            let auth_encoded= authorization.encode();
+
+            // 1. Verify signatures.
+            for si_with_auth in additional_keys.iter() {
+                let si = &si_with_auth.signing_item;
+
+                // Get account_id from signer
+                let account_id_found = match si.signer {
+                    Signer::Key(ref key) =>  Public::try_from(key.as_slice()).ok(),
+                    Signer::Identity(ref id) if <DidRecords>::exists(id) => {
+                        let master_key = <DidRecords>::get(id).master_key;
+                        Public::try_from( master_key.as_slice()).ok()
+                    },
+                    _ => None
+                };
+
+                if let Some(account_id) = account_id_found {
+                    if let Signer::Key(ref key) = si.signer {
+                        // 1.1. Constraint 1-to-1 account to DID
+                        ensure!( Self::can_key_be_linked_to_did( key, si.signer_type),
+                        "One signing key can only belong to one identity");
+                    }
+
+                    // 1.2. Offchain authorization is not revoked explicitly.
+                    ensure!( Self::is_offchain_authorization_revoked((si.signer.clone(), authorization.clone())) == false,
+                        "Authorization has been explicitly revoked");
+
+                    // 1.3. Verify the signature.
+                    let signature = AnySignature::from( Signature::from_h512(si_with_auth.auth_signature));
+                    ensure!( signature.verify( auth_encoded.as_slice(), &account_id),
```

Comment:
Here we are verifying whether the signed data is signed by the respective signer or not, Identity owner can cheat the signing_keys by using any signed data provided by them. We need a process where we can also verify the data of the signature & this process will also allow us to avoid the replay attacks of those signatures.

Reply:
`auth_encoded` contains the data that we are verifying.